test/helpers/sockets.rb

```diff
@@ -0,0 +1,171 @@
+require 'socket'
+
+module TestPuma
+
+  # @!macro [new] req
+  #   @param [String] path request uri path
+  #   @param [Float] dly: delay in app when using 'ci' rackups
+  #   @param [Integer, String] body_conf: response body type and size in kB when using 'ci' rackups
+
+  # @!macro [new] tout
+  #   @param [Float] timeout: read timeout for socket
+
+  # @!macro [new] ret_skt
+  #   @return [SktSSL, SktTCP, SktUNIX] the opened socket
+
+
+  RESP_READ_TIMEOUT = 10
+
+  module SktPrepend
+
+    RESP_READ_LEN = 1_024 * 64
+    RESP_SPLIT = "\r\n\r\n"
+
+    WRITE_TIMEOUT = 5
+
+    # Writes a string to the socket using `syswrite`.
+    # @param [String] str
+    # @return [Integer] number of bytes written
+    def write(str)
+      fast_write str
+    end
+
+    # Writes a string to the socket using `syswrite`.
+    # @param [String] str
+    # @return [self]
+    def <<(str)
+      fast_write str
+      self
+    end
+
+    # Writes a string to the socket using `syswrite`.
+    # @param [String] str the string to write
+    # @return [Integer] the number of bytes written
+    def fast_write(str)
+      return unless str.is_a? String
+      n = 0
+      byte_size = str.bytesize
+      while n < byte_size
+        begin
+          n += syswrite(n.zero? ? str : str.byteslice(n..-1))
+        rescue Errno::EAGAIN, Errno::EWOULDBLOCK
+          unless wait_writable WRITE_TIMEOUT
+            raise ConnectionError, SKT_WRITE_ERR_MSG
+          end
+          retry
+        rescue  Errno::EPIPE, SystemCallError, IOError
+          raise ConnectionError, SKT_WRITE_ERR_MSG
```

Comment:
The definition of `SKT_WRITE_ERR_MSG` seems to be missing

Reply:
Fixed.  I think `ConnectionError` was also a mistake.  Some of this code came from a larger file, which had code for generating a stream of clients, similar to what's in `integration.rb`'s `hot_restart_does_not_drop_connections`.  Sorry for the error.